solr/core/src/java/org/apache/solr/util/ModuleUtils.java

```diff
@@ -0,0 +1,124 @@
+ * Then resolves the lib folder for each, so they can be added to class path.
+ */
+public class ModuleUtils {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final String MODULES_FOLDER_NAME = "modules";
+  private static final Pattern validModNamesPattern = Pattern.compile("[\\w\\d-_]+");
+
+  /**
+   * Returns a path to a module's lib folder
+   * @param moduleName name of module
+   * @return the path to the module's lib folder
+   */
+  public static Path getModuleLibPath(Path solrInstallDirPath, String moduleName) {
+    return getModulesPath(solrInstallDirPath).resolve(moduleName).resolve("lib");
+  }
+
+  /**
+   * Finds list of module names requested by system property or environment variable
+   * @return set of raw volume names from sysprop and/or env.var
+   */
+  static Set<String> resolveFromSyspropOrEnv() {
+    // Fall back to sysprop and env.var if nothing configured through solr.xml
+    Set<String> mods = new HashSet<>();
+    String modulesFromProps = System.getProperty("solr.modules");
+    if (!StringUtils.isEmpty(modulesFromProps)) {
+      mods.addAll(StrUtils.splitSmart(modulesFromProps, ',', true));
+    }
+    String modulesFromEnv = System.getenv("SOLR_MODULES");
+    if (!StringUtils.isEmpty(modulesFromEnv)) {
+      mods.addAll(StrUtils.splitSmart(modulesFromEnv, ',', true));
+    }
+    return mods.stream().map(String::trim).collect(Collectors.toSet());
+  }
+
+  /**
+   * Returns true if a module name is valid and exists in the system
+   */
+  public static boolean moduleExists(Path solrInstallDirPath, String moduleName) {
+    if (!isValidName(moduleName)) return false;
+    Path modPath = getModulesPath(solrInstallDirPath).resolve(moduleName);
+    return Files.isDirectory(modPath);
+  }
+
+  /**
+   * Returns nam of all existing modules
+   */
+  public static Set<String> listAvailableModules(Path solrInstallDirPath) {
+    try (var moduleFilesStream = Files.list(getModulesPath(solrInstallDirPath))) {
+      return moduleFilesStream.filter(Files::isDirectory)
+          .map(p -> p.getFileName().toString()).collect(Collectors.toSet());
+    } catch (IOException e) {
+      log.warn("Found no modules in {}", getModulesPath(solrInstallDirPath), e);
+      return Collections.emptySet();
+    }
+  }
+
+  /**
+   * Parses comma separated string of module names, in practice found in solr.xml.
+   * If input string is empty (nothing configured) or null (e.g. tag not present in solr.xml),
+   * we continue to resolve from system property <code>-Dsolr.modules</code> and if still empty,
+   * fall back to environment variable <code>SOLR_MODULES</code>.
+   * @param modulesFromString raw string of comma-separated module names
+   * @return a set of module
+   */
+  public static Set<String> resolveModulesFromStringOrSyspropOrEnv(String modulesFromString) {
+    Set<String> moduleNames;
```

Comment:
Might as well make this a `Collection<String>` if you are going to `.stream()` it later. Then you don't have to create a new `HashSet` from the list below. Not a big deal though. This is never going to be a huge list...

Reply:
Good idea, fixed.